### DISCUSSION TITLE
Calyx in 2024

### DISCUSSION THREAD

**Discussion by rachitnigam** [cucapra/futil#1825] [General]
Collecting topics for discussion on "Calyx in 2024" meeting.

- Building a performance dashboard for Calyx and the [new Calyx benchmarks repository](https://github.com/calyxir/benchmarks).
- Organizing a new team focused improving Calyx's optimization passes.
  - A specific, large-scale effort is implementing better heuristics for FSM allocation and generation (something @andrewb1999 brought up as a key limitation of existing HLS tools)
    - We can go one step further and perform feedback-directed optimization (FDO) for the FSM allocation pass. Allocate FSMs using some heuristics, run synthesis and get feedback on which FSM nets are on the critical path, reallocate those and try again. This is somewhat similar in spirit to how XLS does FDO for repipelining designs.

**Comment by sampsyo** [comment]
Here are a few categories that occurred to me when we last talked as a group…

### technical debt

This is not necessarily something we want to pour a bunch of effort into actually _doing_ something about, but I would be interested to think about how to at least organize/prioritize various high-order chunks of technical debt. Things on my mind (surely an incomplete list of the important stuff) include:

* switching to a flat/indexed IR (cf #1183)
* stability issues and `with` (cf #1732, #1700, #1699, and more I'm sure)
* exploring a CFG representation (cf #1366, #1225)
* building up more synchronization on/in addition to `@sync`

These are all big efforts with little externally visible reward (well, maybe not the last one). I don't exactly know how, but I think we should keep them in mind anyway, even if that "keeping in mind" is essentially coming up with a clear description of how much work would be involved (and hence a reason not to do it).

### launching our new open-source organization

We have taken the first step toward "legitimizing" Calyx as an open-source project with the new `calyxir` org. Let's think about the work required to do the other stuff we've talked about:

* Zulip or whatever (#1803)
* advertise weekly meetings externally
* formalizing an RFC process
* assigning official roles
